### PR TITLE
[MAD][DATA1018][Lab sklearn and unsupervised lear] Elena jP

### PR DESCRIPTION
**Pull request by elenajpp**
Belen lo he hecho pero en la parte final del challege tres me no me lo resuelve por que dice que esta en float y no se como solucionarlo 
Un saludo 
Elena

### CODE REVIEW

File: module-3/lab-sklearn-and-unsupervised-learning/your-code/main.ipynb [hunk 1]
```diff
@@ -12,7 +12,7 @@
   },
```
Comment (Gobinde43) [inline module-3/lab-sklearn-and-unsupervised-learning/your-code/main.ipynb]: - In this one, "What is the average ash content for each cluster?". try to do something like the following:
```Python
(np.average(wine['data'][:,2][np.where(kmeans.labels_ == 0)]), np.average(wine['data'][:,2][np.where(kmeans.labels_ == 1)]), 
np.average(wine['data'][:,2][np.where(kmeans.labels_ == 2)]), np.average(wine['data'][:,2][np.where(kmeans.labels_ == 3)]))
```
- "For the prescribed_medicines column, fill the missing values with the value no prescription. For the diagnosis column, fill the missing values with no diagnosis." better in one line when possible and the code is readable enough:
```Python
patients.fillna({'prescribed_medicines': 'no prescription', 'diagnosis': 'no diagnosis'}, inplace=True)
````
- In this one, "How about doctor_name? Since a doctor visit without a doctor name might not be meaningful, we will drop these rows", you don't need to specify either axis or how parameters:
```Python
# your code 
patients.dropna(axis=0, how='any', inplace=True)
# without parameters 
patients.dropna(inplace=True)
```
- Good to use the label encoder. Other way of doing it:
```Python
# your code
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

patients['patient_diabetic'] = le.fit_transform(patients['patient_diabetic'])
patients['patient_allergic'] = le.fit_transform(patients['patient_allergic'])
patients['patient_show'] = le.fit_transform(patients['patient_show'])
patients['is_regular_visit'] = le.fit_transform(patients['is_regular_visit'])
display(patients.head())
# Other way
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()
label.fit([False, True])
patients[['patient_diabetic', 'patient_allergic', 'patient_show', 'is_regular_visit']] = patients[['patient_diabetic', 'patient_allergic', 'patient_show', 'is_regular_visit']].apply(label.fit_transform)
patients.head()
```
- In this one "The last step is to handle the object data." You can simplify your code in the following way:
```Python
# your code 
for i in patients.columns:
    if patients[i].dtype == object: 
        print (patients[i].value_counts())
# other way 
print(patients.patient_gender.value_counts())
print('-----')
print(patients.doctor_name.value_counts())
print('-----')
print(patients.prescribed_medicines)
print('-----')
print(patients.diagnosis.value_counts())
```
- In this one, "In the next cells, do the following:
Create a new column called diagnosis_int that has 0 and 1 based on the values in diagnosis.
Create a new column called patient_gender_int that has 0 and 1 based on the values in patient_gender.
Drop the following columns: doctor_name, diagnosis, prescribed_medicines, and patient_gender."  do the following to avoid the execution error in drop statement: 
```Python
patients['diagnosis_int'] = np.where(patients.diagnosis == "no diagnosis", 0, 1)
patients['patient_gender_int'] = np.where(patients.patient_gender == "Male", 0, 1)
patients.drop(columns=['doctor_name', 'diagnosis', 'prescribed_medicines', 'patient_gender'], inplace=True)
```
- Try to complete the lab when you have some free time. 
- We can divide (as a taxonomy) the clustering algorithms in two classes:
1. Clustering based in distance: observations can be represented as points in a dimensional space where the dimensions are the features we have. The distance metric can be (euclidean, manhattan, mahalanobis) to separate the different clusters. K-means belongs to this class and it's the most famous one.
2. Clustering based on density: this algorithms don't into account only the distance among "points" but also detecting the areas where the density of points is higher.
- Find below a link to the main clustering algorithms:
https://towardsdatascience.com/the-5-clustering-algorithms-data-scientists-need-to-know-a36d136ef68
- Sklearn has all this clustering algorithms:
https://scikit-learn.org/stable/modules/classes.html#module-sklearn.cluster
